WalletWasabi/CoinJoin/Coordinator/Rounds/CoordinatorRound.cs

```diff
@@ -1167,46 +1167,83 @@ public void AddBob(Bob bob)
 
 		public async Task<IEnumerable<Alice>> RemoveAlicesIfAnInputRefusedByMempoolAsync()
 		{
-			var alicesRemoved = new List<Alice>();
-			var key = new Key();
-
 			using (await RoundSynchronizerLock.LockAsync().ConfigureAwait(false))
 			{
+				// Removing Alice without penalty can only happen in early phases.
 				if ((Phase != RoundPhase.InputRegistration && Phase != RoundPhase.ConnectionConfirmation) || Status != CoordinatorRoundStatus.Running)
 				{
 					throw new InvalidOperationException("Removing Alice is only allowed in InputRegistration and ConnectionConfirmation phases.");
 				}
 
-				var checkingTasks = new List<(Alice alice, Task<(bool accepted, string reason)> task)>();
-				var batch = RpcClient.PrepareBatch();
-				foreach (Alice alice in Alices)
+				// If we can build a transaction that the mempool accepts, then we're good, no need to remove any Alices.
+				(bool accept, string rejectReason) resultAll = await TestMempoolAcceptWithTransactionSimulationAsync().ConfigureAwait(false);
+				if (resultAll.accept)
 				{
-					// Check if mempool would accept a fake transaction created with the registered inputs.
-					// This will catch ascendant/descendant count and size limits for example.
-					checkingTasks.Add((alice, batch.TestMempoolAcceptAsync(alice.Inputs)));
+					return Enumerable.Empty<Alice>();
 				}
-				var waiting = Task.WhenAll(checkingTasks.Select(t => t.task));
-				await batch.SendBatchAsync().ConfigureAwait(false);
-				await waiting.ConfigureAwait(false);
+				Logger.LogInfo($"Mempool acceptance is unsuccessful! Number of Alices: {Alices.Count}.");
 
-				foreach (var t in checkingTasks)
+				var alicesSpent = new HashSet<Alice>();
+				var alicesUnconfirmed = new HashSet<Alice>();
+				// The created tx was not accepted. Let's figure out why. Is it because an Alice doublespent or because of too long mempool chains.
+				foreach (var alice in Alices)
 				{
-					var result = await t.task.ConfigureAwait(false);
-					if (!result.accepted)
+					foreach (var input in alice.Inputs.Select(x => x.Outpoint))
 					{
-						alicesRemoved.Add(t.alice);
-						Alices.Remove(t.alice);
-						Logger.LogInfo($"Mempool acceptance failed: {result.reason ?? "no reason"}.");
+						var response = await RpcClient.GetTxOutAsync(input.Hash, (int)input.N, includeMempool: true).ConfigureAwait(false);
+						if (response is null)
+						{
+							alicesSpent.Add(alice);
```

Comment:
We should `break` here because it makes no sense to continue the rest of the Alices' utxos.

Reply:
Do you mean Alice's?